packages/sourcecred/src/core/clearcred/scoredContribution.js

```diff
@@ -0,0 +1,140 @@
+  +score: number,
+|};
+
+export type ScoredContribution = {|
+  +id: string,
+  +plugin: string,
+  +type: string,
+  +timestampMs: TimestampMs,
+  +expression: ScoredExpression,
+  +participants: $ReadOnlyArray<{|
+    +id: string,
+    +shares: $ReadOnlyArray<{|
+      +amount: number,
+      +key: string,
+    |}>,
+    +score: number,
+  |}>,
+|};
+
+const scoreExpression: (Expression, Config) => ScoredExpression = (
+  expression,
+  config
+) => {
+  const scoredExpressionOperands = expression.expressionOperands.map(
+    (operand) => scoreExpression(operand, config)
+  );
+  const scoredWeightOperands = expression.weightOperands.map((operand) => ({
+    ...operand,
+    score: getWeight(operand, config),
+  }));
+  let score;
+  const operatorOptional = OPERATORS.find((o) => o === expression.operator);
+  if (operatorOptional) {
+    score = applyOperator(
+      operatorOptional,
+      scoredWeightOperands,
+      scoredExpressionOperands,
+      config
+    );
+  } else {
+    const operatorKey = expression.operator;
+    const expressionWithParsedOperator = {
+      ...expression,
+      operator: getOperator(operatorKey, config),
+    };
+    return {
+      ...scoreExpression(expressionWithParsedOperator, config),
+      operatorKey,
+    };
+  }
+  if (!scoredExpressionOperands.length && !scoredWeightOperands.length) {
+    score = 0;
+  }
+  return {
+    operator: operatorOptional,
+    description: expression.description,
+    score,
+    expressionOperands: scoredExpressionOperands,
+    weightOperands: scoredWeightOperands,
+  };
+};
+
+export const scoreContribution = (
+  contribution: Contribution,
+  config: Config
+): ScoredContribution => {
+  const scoredExpression = scoreExpression(contribution.expression, config);
+  let totalShares = 0;
+  const participants = contribution.participants.map((participant) => {
+    const participantShares = participant.shares.map((share) => ({
+      ...share,
+      amount: getShares(share.key, config),
+    }));
+    const sumParticipantShares = participantShares
+      .map((s) => s.amount)
+      .reduce((a, b) => a + b, 0);
+    totalShares += sumParticipantShares;
+    return {
+      ...participant,
+      shares: participantShares,
+      score: (scoredExpression.score * sumParticipantShares) / totalShares,
+    };
+  });
+  return {
+    ...contribution,
+    participants,
+    expression: scoredExpression,
+  };
+};
+
+export function scoreContributions(
+  contributions: Iterable<Contribution>,
+  configs: $ReadOnlyArray<Config>
+): Iterable<ScoredContribution> {
+  return (function* () {
+    const orderedConfigs = configs
+      .slice()
+      .sort((a, b) => a.startTimeMs - b.startTimeMs);
+    for (const contribution of contributions) {
+      const applicableConfig: Config | void = findLast(
+        orderedConfigs,
+        (config) => config.timestampMs < contribution.timestampMs
+      );
+      if (!applicableConfig) continue;
```

Comment:
Given that this is a core algorithm calculation, why not throw an error here? In what circumstance is not having an applicable config an allowable scenario?

Reply:
This makes it especially easy to totally omit contributions before a certain date, which is a request we get particularly for forked repos. It seems nicer than making users create a zeroed-out config with a 0 or -Infinity timestamp.